nova_vm/src/ecmascript/builtins/numbers_and_dates/math_object.rs

```diff
@@ -377,87 +475,332 @@ impl MathObject {
+            return Ok(n.into_value());
+        }
+
+        // 4. Let n32 be the result of converting n to IEEE 754-2019 binary32 format using roundTiesToEven mode.
+        let n32 = n.into_f32(agent).round_ties_even();
+
+        // 5. Let n64 be the result of converting n32 to IEEE 754-2019 binary64 format.
+        let n64 = n32 as f64;
+
+        // 6. Return the ECMAScript Number value corresponding to n64.
+        Ok(Value::from_f64(agent, n64))
     }
 
-    fn hypot(_agent: &mut Agent, _this_value: Value, _arguments: ArgumentsList) -> JsResult<Value> {
-        todo!();
+    fn hypot(agent: &mut Agent, _this_value: Value, arguments: ArgumentsList) -> JsResult<Value> {
+        // 1. Let coerced be a new empty List.
+        let mut coerced = vec![];
+
+        // 2. For each element arg of args, do
+        for &arg in arguments.iter() {
+            // a. Let n be ? ToNumber(arg).
+            let n = to_number(agent, arg)?;
+
+            // b. Append n to coerced.
+            coerced.push(n);
+        }
+
+        // 3. For each element number of coerced, do
+        for number in coerced.iter() {
+            // a. If number is either +∞𝔽 or -∞𝔽, return +∞𝔽.
+            if number.is_pos_infinity(agent) || number.is_neg_infinity(agent) {
+                return Ok(Value::infinity());
+            }
+        }
+        
+        // 4. Let onlyZero be true.
+        let mut only_zero = true;
+
+        // 5. For each element number of coerced, do
+        for number in coerced.iter() {
+            // a. If number is NaN, return NaN.
+            if number.is_nan(agent) {
+                return Ok(Value::nan());
+            }
+
+            // b. If number is neither +0𝔽 nor -0𝔽, set onlyZero to false.
+            if !number.is_pos_zero(agent) && !number.is_neg_zero(agent) {
+                only_zero = false;
+            }
+        }
+
+        // 6. If onlyZero is true, return +0𝔽.
+        if only_zero {
+            return Ok(Value::zero());
+        }
+
+        // 7. Return an implementation-approximated Number value representing the square root of the sum of squares of the mathematical values of the elements of coerced.
+        return Ok(Value::from_f64(agent, coerced.iter().map(|n| n.into_f64(agent)).fold(0.0, |acc, n| acc + n * n).sqrt()));
     }
 
     fn imul(agent: &mut Agent, _this_value: Value, arguments: ArgumentsList) -> JsResult<Value> {
-        let _x = to_number(agent, arguments.get(0))?;
-        todo!();
+        // 1. Let a be ℝ(? ToUint32(x)).
+        let a = to_uint32(agent, arguments.get(0))?;
+
+        // 2. Let b be ℝ(? ToUint32(y)).
+        let b = to_uint32(agent, arguments.get(1))?;
+        
+        // 3. Let product be (a × b) modulo 2**32.
+        let product = (a * b) % 2u32.pow(32);
+
+        // 4. If product ≥ 2**31, return 𝔽(product - 2**32); otherwise return 𝔽(product).
+        if product >= 2u32.pow(31) {
+            Ok(Value::from(product as i32 - 2i32.pow(32)))
```

Comment:
**issue:** And this overflows. Actually, this 4. step is seemingly just equal to `product as i32`.